### PR TITLE
Upgrade to RN74

### PR DESCRIPTION
**Pull request by gosha212**
## Description

- This pull request addresses the issue described here: #4484

In this pull request, I have added support for rn74

---

> _For features/enhancements:_
 - [ ] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [ ] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

### CODE REVIEW

File: .buildkite/pipeline_common.sh [hunk 1]
```diff
@@ -2,11 +2,14 @@
 
 echo "steps:"
 
+cat .buildkite/jobs/pipeline.ios_rn_74.yml
```
Comment (d4vidi) [inline .buildkite/pipeline_common.sh:5]: That's a hefty bulk-load of agents running concurrently. We defn shouldn't get rid of 73, but perhaps its time to get rid of 71, anyways.

File: detox/android/build.gradle [hunk 2]
```diff
@@ -3,12 +3,12 @@ buildscript {
 
     ext {
         isOfficialDetoxLib = true
-        kotlinVersion = '1.8.0'
+        kotlinVersion = '1.9.20'
```
Comment (d4vidi) [inline detox/android/build.gradle:6]: That's the new minimum-required version?

File: README.md [hunk 3]
```diff
@@ -60,7 +60,7 @@ The most difficult part of automated testing on mobile is the tip of the testing
 
 Detox was built from the ground up to support React Native projects.
 
-While Detox should work out of the box with almost any React Native version of the latest minor releases, official support is provided for React Native versions `0.71.x`, `0.72.x` and `0.73.x` without React Native's ["New Architecture"](https://reactnative.dev/docs/the-new-architecture/landing-page).
+While Detox should work out of the box with almost any React Native version of the latest minor releases, official support is provided for React Native versions `0.71.x`, `0.72.x`, `0.73.x`, `0.74.x` without React Native's ["New Architecture"](https://reactnative.dev/docs/the-new-architecture/landing-page).
```
Comment (d4vidi) [inline README.md:63]: 🆒

File: detox/test/ios/example/PrivacyInfo.xcprivacy [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (d4vidi) [inline detox/test/ios/example/PrivacyInfo.xcprivacy:1]: What the heck is this?

File: detox/test/package.json [hunk 5]
```diff
@@ -50,16 +50,18 @@
...
-    "@react-native/typescript-config": "0.73.1",
+    "@react-native/babel-preset": "0.74.88",
+    "@react-native/eslint-config": "0.74.88",
+    "@react-native/metro-config": "0.74.88",
+    "@react-native/typescript-config": "0.74.88",
     "@tsconfig/react-native": "^3.0.0",
     "@types/jest": "^29.5.11",
     "@types/node": "^16.18.68",
-    "@types/react": "^18.2.45",
+    "@types/react": "^18.2.6",
```
Comment (d4vidi) [inline detox/test/package.json:60]: Why the version downgrade?

File: detox/test/package.json [hunk 6]
```diff
@@ -50,16 +50,18 @@
...
     "@typescript-eslint/eslint-plugin": "^6.16.0",
     "@typescript-eslint/parser": "^6.16.0",
     "axios": "^1.7.7",
+    "babel-jest": "^29.6.3",
```
Comment (d4vidi) [inline detox/test/package.json:64]: Is this really needed (my guess is that - no)

File: examples/demo-react-native/package.json [hunk 7]
```diff
@@ -24,17 +24,19 @@
...
+    "@babel/preset-env": "^7.20.0",
+    "@react-native/babel-preset": "0.74.88",
+    "@react-native/eslint-config": "0.74.88",
+    "@react-native/metro-config": "0.74.88",
+    "@react-native/typescript-config": "0.74.88",
     "@tsconfig/react-native": "^2.0.2",
     "@types/fs-extra": "^9.0.13",
     "@types/jest": "^29.2.1",
-    "@types/react": "^18.0.24",
+    "@types/react": "^18.2.6",
```
Comment (d4vidi) [inline examples/demo-react-native/package.json:39]: Version downgrade #2

File: examples/demo-react-native/android/gradle.properties [hunk 8]
```diff
@@ -9,7 +9,7 @@
 
 # Specifies the JVM arguments used for the daemon process.
 # The setting is particularly useful for tweaking memory settings.
-org.gradle.jvmargs=-Xmx2048m -Dfile.encoding=UTF-8
+org.gradle.jvmargs=-Xmx3g -Dfile.encoding=UTF-8
```
Comment (d4vidi) [inline examples/demo-react-native/android/gradle.properties:12]: these guys are unbelievable

File: detox/test/package.json [hunk 9]
```diff
@@ -50,16 +50,18 @@
     "@babel/core": "^7.20.0",
     "@babel/preset-env": "^7.20.0",
     "@babel/runtime": "^7.20.0",
-    "@react-native/eslint-config": "^0.73.2",
-    "@react-native/metro-config": "^0.73.3",
-    "@react-native/typescript-config": "0.73.1",
+    "@react-native/babel-preset": "0.74.88",
+    "@react-native/eslint-config": "0.74.88",
+    "@react-native/metro-config": "0.74.88",
+    "@react-native/typescript-config": "0.74.88",
```
Comment (d4vidi) [inline detox/test/package.json:56]: Is this really needed?

File: examples/demo-react-native/package.json [hunk 10]
```diff
@@ -24,17 +24,19 @@
...
   },
   "devDependencies": {
-    "@babel/preset-react": "^7.23.3",
-    "@react-native/babel-preset": "0.73.19",
-    "@react-native/metro-config": "^0.73.2",
+    "@babel/preset-env": "^7.20.0",
+    "@react-native/babel-preset": "0.74.88",
+    "@react-native/eslint-config": "0.74.88",
+    "@react-native/metro-config": "0.74.88",
+    "@react-native/typescript-config": "0.74.88",
```
Comment (d4vidi) [inline examples/demo-react-native/package.json:35]: Is this really needed? #2

**Review by d4vidi — COMMENTED** [review]
Great to see such a lightweight upgrade, for a change. See some comments below...

File: .buildkite/pipeline_common.sh (hunk 1 above)
Comment (gosha212) [reply .buildkite/pipeline_common.sh:5]: Removed

File: detox/android/build.gradle (hunk 2 above)
Comment (gosha212) [reply detox/android/build.gradle:6]: No, this is what the template wants

File: detox/test/package.json (hunk 9 above)
Comment (gosha212) [reply detox/test/package.json:56]: We need to align the versions

File: detox/test/package.json (hunk 6 above)
Comment (gosha212) [reply detox/test/package.json:64]: Copy pasted it from the template.

File: examples/demo-react-native/android/gradle.properties (hunk 8 above)
Comment (gosha212) [reply examples/demo-react-native/android/gradle.properties:12]: Yes. we had OOM during the build

**Review by d4vidi — APPROVED** [review]

**Pull request merged**